src/AppBundle/Controller/TqirAppApiController.php

```diff
@@ -134,12 +142,83 @@ private function registerAction(Request $request, $notificationType, $notificati
 
     /**
      * @param Request $request
-     * @param $notificationType
-     * @param $notificationAddress
+     * @param string $notificationType
+     * @param string $notificationAddress
      *
-     * @return mixed
+     * @return Response
+     * @throws \AppBundle\Tiqr\Exception\ConfigurationException
      */
     private function loginAction(Request $request, $notificationType, $notificationAddress)
     {
+        $now = new \DateTimeImmutable();
+        $sessionKey = $request->get('sessionKey');
+        $userId = $request->get('userId');
+        $response = $request->get('response');
+
+        try {
+            $user = $this->userRepository->getUser($userId);
+        } catch (UserNotExistsException $e) {
+            return new Response('INVALID_USER', Response::HTTP_BAD_REQUEST);
+        }
+
+        // Check if user is blocked. (TODO: maybe do this in AuthenticationController?)
+        if ($this->configuration->temporaryBlockEnabled() &&
+            $user->isBlockTemporary($now, $this->configuration->getTemporaryBlockDuration())) {
+            return new Response('ACCOUNT_BLOCKED', Response::HTTP_FORBIDDEN);
+        }
+
+        if (!$this->configuration->temporaryBlockEnabled() && $user->isBlocked()) {
+            return new Response('ACCOUNT_BLOCKED', Response::HTTP_FORBIDDEN);
+        }
+
+        // Verify the app's response.
+        $result = $this->tiqrService->authenticate($user, $response, $sessionKey);
+        if ($result->isValid()) {
+            $user->resetLoginAttempts();
+            $user->updateNotification($notificationType, $notificationAddress);
+
+            return new Response($result->getMessage(), Response::HTTP_OK);
+        }
+
+        // The user did something wrong.
+        if ($result instanceof RejectedAuthenticationResponse) {
+            // If there is no limit how many times the user can try to login.
+            if (!$this->configuration->hasMaxLoginAttempts()) {
+                return new Response($result->getMessage(), Response::HTTP_FORBIDDEN);
+            }
+
+            // Does the user still have login attempts left?.
+            if ($user->getLoginAttempts() < ($this->configuration->getMaxAttempts() - 1)) {
```

Comment:
This logical expression in this statement is what you talked about during the dev chat, right? Would it be beneficial to move some of these user/configuration checks to a separate class? In the previous PR we already discussed this as an option. Just thinking out loud here.

Reply:
An even simpler solution could be to move these more logic heavy assertions to the `TiqrUser` class.